src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -409,9 +401,81 @@ def log_zinc_file(self, analysis_file):
                                    if os.path.exists(analysis_file)
                                    else 'nonexistent'))
 
+  @classmethod
+  def _javac_plugin_args(cls, javac_plugin_map):
+    ret = []
+    for plugin, args in javac_plugin_map.items():
+      for arg in args:
+        if ' ' in arg:
+          # Note: Args are separated by spaces, and there is no way to escape embedded spaces, as
+          # javac's Main does a simple split on these strings.
+          raise TaskError('javac plugin args must not contain spaces '
+                          '(arg {} for plugin {})'.format(arg, plugin))
+      ret.append('-C-Xplugin:{} {}'.format(plugin, ' '.join(args)))
+    return ret
+
+  @classmethod
+  def _scalac_plugin_args(cls, scalac_plugin_map, classpath):
+    if not scalac_plugin_map:
+      return []
+
+    plugin_jar_map = cls._find_scalac_plugins(scalac_plugin_map.keys(), classpath)
+    ret = []
+    for name, jar in plugin_jar_map.items():
+      ret.append('-S-Xplugin:{}'.format(jar))
+      for arg in scalac_plugin_map[name]:
+        ret.append('-S-P:{}:{}'.format(name, arg))
+    return ret
+
+  @staticmethod
+  def _find_scalac_plugins(scalac_plugins, classpath):
+    """Returns a map from plugin name to plugin jar/dir."""
+    # Allow multiple flags and also comma-separated values in a single flag.
+    plugin_names = set([p for val in scalac_plugins for p in val.split(',')])
+    active_plugins = {}
+    buildroot = get_buildroot()
+
+    def process_info_file(classpath_element, plugin_info_file):
+      plugin_info = ElementTree.parse(plugin_info_file).getroot()
+      if plugin_info.tag != 'plugin':
+        raise TaskError(
+          'File {} in {} is not a valid scalac plugin descriptor'.format(
+            _SCALAC_PLUGIN_INFO_FILE, classpath_element))
+      name = plugin_info.find('name').text
+      if name in plugin_names:
+        # It's important to use relative paths, as the compiler flags get embedded in the zinc
+        # analysis file, and we port those between systems via the artifact cache.
+        rel_classpath_element = os.path.relpath(classpath_element, buildroot)
+        # Some classpath elements may be repeated, so we allow for that here.
+        if active_plugins.get(name, rel_classpath_element) != rel_classpath_element:
+          raise TaskError('Plugin {} defined in {} and in {}'.format(name, active_plugins[name],
+                                                                     classpath_element))
+        active_plugins[name] = rel_classpath_element
+
+    for c in classpath:
```

Comment:
Since `_find_scalac_plugins` will run for every compilation, it would be good to memoize some of this. Maybe extract the body of this loop (each classpath entry) into a `memoized_method`?